### DISCUSSION TITLE
Options for Metadata storage/consumption

### DISCUSSION THREAD

**Discussion by jjspierx** [immich-app/immich#5530] [General]
I thought I would start a discussion about metadata storage.  The different ways metadata for images/videos can be stored and the pros/cons of each storage method.  I would also like to discuss the various ways that Immich consumes metadata and some ideas for changes in the future.  I am a developer, however, I have not done any development work on Immich, so a lot of what I say is assumptions.  Please correct me if I am wrong on anything.

As I understand it, there are essentially 3 ways to store metadata with respect to Immich:

1. Immich database
2. EXIF data
3. XMP sidecar data

**Immich Database**

I assume that when files are uploaded into Immich, Immich looks at exif data and XMP sidecar data, and adds the pertinent information for Immich into it's database.  This is probably by far the fastest method of extracting data about media as it doesn't even need a file system in order to obtain information about media, and db operations are much faster than file system operations.

Pros:
- fast
- can make changes without altering original file

Cons:
- Requires a relationship with a file in the file system that could be broken if a filename/path changes and db is not updated to reflect this (could be caused by bad coding, uncaught exception, etc...)
- If db is corrupted, deleted, or lost in anyway, meta-data not stored on the file system will be lost

**EXIF data**

EXIF data I assume is where Immich gets most of the metadata that it needs for it's database.  I am guessing it is mostly only used on the the initial upload into Immich.

Pros:
- Directly attached to the media it belongs to
- Exists even if Immich db is lost and can be used to recreate a db with Immich or another platform

Cons:
- Slow access 

**XMP sidecar**

XMP sidecar metadata is used to supplement EXIF data.  I do not know how Immich uses XMP sidecar metadata.  I would guess it uses it just like EXIF data and is extracted on upload and added to the Immich db.  Or maybe sidecar data is kept out of the db and read on demand from the file system?

Pros:
- Allows storage/modification of metadata without affecting the original media file 

Cons:
- Slow access
- Not directly attached to media, so if link to media is broken the metadata is lost (if filename of media or media path is changed and sidecar is not updated to reflect this for example).

**Discussion**

Now I would like to discuss how Immich uses and modifies metadata.  As I understand it, Immich as well as most if not all photo storage systems prefer to not modify the original media when possible.  So for instance, if I modify the photo taken time, GPS coordinates, description, favorite, albums, or any other metadata, the db is updated to reflect this, but the original file is not modified.  I would guess the reasons for this are:

1. Non-destructive: Original file is maintained, treated like an unalterable negative. Can always revert back to the original image if some "change" goes wrong.
2. Safe: Modifying a file directly has inherent risk of corruption, deletion, or even user error (accidentally changed metadata on wrong file for instance).
3.  Ease of coding: Much easier to update a db entry alone, then to also modify a file EXIF data or create/edit a sidecar.

However, there are potential problems with ONLY modifying metadata in the DB.  I am thinking long-term.  What happens if I have a hard-drive failure and Immich has not been backed up?  What happens if I decide to switch from Immich to another solution in 10 years?  I would like for any changes I have made to metadata to be stored outside of Immich.  This could be either directly to the file, or by creating a sidecar file for the media.

I think a potential non-destructive way to ensure that all metadata is stored and retained in a way that could easily be brought out of Immich or used to recreate an Immich db from scratch, would be to allow the option of mirroring db and sidecar data.  When a file is uploaded into Immich, the EXIF and db metadata should match, and no sidecar would be needed.  But once any metadata is updated (adding a description, tags, people, objects for instance), Immich could update the db metadata, and then create a sidecar for the media to store the edit as well.  This way, there would always be a complete backup of meta changes.  This would make recreating an Immich db much faster, as Immich could skip a lot of the steps such as ML for people and objects since this information would be extractable from the sidecars.

Anyway, just some ideas I have been thinking about.  Please correct me if I am wrong about any of the assumptions made above.

**Comment by bo0tzz** [comment]
Your description is mostly correct. The groundwork to support writing metadata to a sidecar file was recently laid in #5066. Long-term, it would indeed be good to write as much data as we can into those files as well for robustness. See also e.g.  #4348

**Comment by Randy-Blancett** [comment]
Just throwing it out there, I understand why Immich would not want to modify the original by default, however Over the last 5 years I have migrated between several different plat forms Google, Synology, DigiKam. Although having the side cars is better than nothing, It still means that the history of the data is lost if the image it downloaded or moved to a different system. For example, I have scanned thousands of physical photos so that my family can keep the memories, and try to add context to them through the comments on the photos (or even just the face tagging) However when I move the photos over to Immich all that information seems to be lost, In addition even if I start over, when a family member downloads the image all the information will be lost because it is stored in the DB not in the image its self.

**Reply by jrasm91** [reply]
The recommended way to overcome this is for all photo editing software to correctly read and write sidecar files, which is the direction we're pursuing.

**Reply by Randy-Blancett** [reply]
I have to say that I think having the data in two places opens this up for it being lost, however at least the sidecar files gives an opportunity to import the data when a better system comes along. From your comment I assume the behavior that I am currently seeing is a bug or will be "fixed" in future iterations. I am currently seeing where I upload a picture with a description however when it is imported the description is missing. I have tried with both Just the image / embedded metadata and uploading the file with a sidecar. even worse when I upload a sidecar metadata file that has been generated by digiKam the Description is not filled in, If i add a description Immich creates a new sidecar file that has the description but non of the other information that the picture originally had.

**Reply by aisbergde** [reply]
Of course, immich should synchronize content, but not just store and overwrite existing (regarding descriptions or other content).

It is fine, when immich will not write into the original assets. If this is required, other software can do this (I use digiKam).
And it is also fine, when immich for now uses only a limited set of metadate. immich will grow, more metadata will be used and synchronized.

The only important thing is, that synchronizing between database and sidecar files happens at all and that sync is safe. Maybe it is safe, that immich uses their own sidecar files. But is makes is hard to synchronize with other software. I think, it is much better, to use only one common sidecar file per asset. I don't want a new metadata island "immich".

**Comment by kvalev** [comment]
On that note, are there any plans on making the sidecar folder configurable? I would really prefer to have my external library mounted as read-only, but that becomes inconvenient as soon as you need to modify a date or a location.

**Reply by TapuCosmo** [reply]
I actually just finished configuring my external libraries to prevent deleting images while allowing metadata edits. The following is mostly applicable to adding files to external libraries through Samba, but can probably be adapted to other configurations.

## Steps
1. Edit `/etc/samba/smb.conf` and add `inherit permissions = yes` and `force group = <group>` to your share. Ensure that the Immich user is a member of the specified group. Additionally, ensure that files that are uploaded to Samba are not owned by the Immich user.
2. `systemctl restart smbd`
3. `cd` to your external library directory. Make sure you are in the correct directory before continuing, as you can break your system if you `cd` to the wrong directory and recursively change files (especially if you are running these commands as root).
4. If necessary, use `chown -R <user>:<group> .` to recursively fix the external library to have the correct user (not Immich user) and group.
5. Use `find . -type d -exec chmod 3770 -- {} +` (change the last digit to suit your existing needs) to enable the SGID and sticky bits in the external library. The sticky bit will prevent Immich from editing/deleting any files it does not own (your external library images) while allowing it to create and edit the XMP sidecar files under its ownership.
6. Edit your `docker-compose.yml` to mount your external library as `:rw` and restart Immich.

To validate that you set this up correctly, add a test image to your external library and attempt to permanently delete it through Immich (delete it from the trash). Upon rescanning the library for new files, it should reappear in your photos. Also, try editing the metadata.

Alternative solution: `unionfs-fuse` with your external library as the lower and a blank directory as the upper. This breaks fsnotify though (so no automatic library watching), which is why I moved to the above solution.

**Reply by kvalev** [reply]
I finally had the time to look into it and there is another way to achieve mounting your external library as read-only, but allowing for sidecar files to be written - by using an OverlayFS volume.

In docker compose this would like this:

```yaml

services:
  immich-server:
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    volumes:
      ...
      - immich-external-library:/mnt/data/my-external-library
...


volumes:
  ...
  immich-external-library:
    driver: local
    driver_opts:
      type: overlay
      o: lowerdir=PATH_TO_THE_READONLY_EXTERNAL_LIBRARY,upperdir=PATH_TO_THE_WRITEABLE_SIDECAR_FOLDER,workdir=PATH_TO_SOME_RANDOM_FOLDER_WHICH_WILL_BE_USED_AS_WORKDIR_BY_OVERLAYFS
      device: overlay
```

Whatever you specify as `lowerdir` will be mounted as readonly, so you dont have to fear changes to your files. All sidecar files will be stored in `upperdir` and the folder layout will mimick the one in the external library. `workdir` can be some random folder which will be used by OverlayFS. All three folders have to exist before starting the stack and the paths have to be absolute.

Note that this config did not work for me when using the colima container runtime on macOS, but it worked on Debian.

Some references:
https://serverfault.com/questions/841238/how-to-use-overlayfs-with-docker-volumes
https://stackoverflow.com/questions/29550736/can-i-mount-docker-host-directory-as-copy-on-write-overlay

@di-ivanov Maybe this would be of interest for you...

**Reply by vasishche** [reply]
@kvalev It seems that without immich restart, the search for new files does not work (https://github.com/immich-app/immich/issues/10538#issuecomment-2676493200).

@TapuCosmo, there are 2 problems with your method:
- when you create a new subfolder using samba user, sticky bit is not inherited, so all files in the new subfolders can be deleted,
- even in sticky bit folders, files are only protected from deletion, files can still be edited - erased/corrupted in this way.

Thus, both solutions are unacceptable. 😢

**Reply by kvalev** [reply]
The overlay solution is working great for me so far. Whenever new files appear in the external library, they get reindexed without any issues. I am not sure why they have to restart, maybe they messed up their config or something.

**Reply by vasishche** [reply]
Finally, I succeeded with the solution above by simply changing the 1st step to `force directory mode = 1770` and `create mask = 0644` instead of `inherit permissions = yes`.

**Comment by di-ivanov** [comment]
I second the opinion of @kvalev for making the sidecar folder configurable, or at least have a fallback if not writable. I am new to immich, and I want to say kudos for the great piece of software. I second the opinion of @kvalev for making the sidecar folder configurable, or at least have a fallback if not writable.
After a short look at the code and the DB schema it doesn't look like complicated scenario at all. I tested a small PoC by just changing the DB a little. On my test instance I just populated the sidecarPath column by concatenating the originalPath to a writable directory:
```
UPDATE public.assets AS a SET "sidecarPath" = 
	concat('/usr/src/app/sidecars', (SELECT b."originalPath" FROM public.assets b 
	WHERE b.id = id LIMIT 1), '.xmp')
WHERE a."sidecarPath" IS NULL;
```
My test library was flat and I had to create the writable parent folder manually, but after that immich just uses this path smoothly.
My question is, does this clash with the future plans for immich, or is it aligned. I saw a recent PR that allows deletion of assets in read only external library, so to me it seems as a logical next step. If this is fine I can expand my PoC and turn it into working code.

**Reply by jrasm91** [reply]
This is fine with me, but I don't know the best way to expose this feature to the user and how it should be configurable. You almost have the "storage template" feature for sidecar files now. Maybe just the two options for "storing them next to assets" vs "storing them in their own directory" would be sufficient. 

Their own directory could be hard coded to be MEDIA_LOCATION + /sidecars/ + original path. 

The user could mount any folder into /usr/src/app/sidecars

**Reply by zolakt** [reply]
Option for "storing them next to assets" vs "storing them in their own directory" would be great.

My use case is mostly for external libraries. I share these with other people that don't use Immich (or any other similar solution), they just browse pictures through the filesystem. The files structure would look cleaner without all the .xmp files

**Reply by ned-martin** [reply]
> Option for "storing them next to assets" vs "storing them in their own directory" would be great.
> 
> My use case is mostly for external libraries. I share these with other people that don't use Immich (or any other similar solution), they just browse pictures through the filesystem. The files structure would look cleaner without all the .xmp files

Note that you can easily automate merging the XMP sidecar files into your image files if you want to clean up your filesystem:

`exiftool -tagsfromfile SRC.xmp -xmp DST.jpg`

**Reply by SinTan1729** [reply]
This will be great. I'm migrating from PhotoPrism, and loving it so far. I merged the corrected date etc. to the original files, that's fine. But my library is heavily tagged, and I'd love to be able to tag theme here too, while keeping in a read-only external libary. (I have a migration plan, but this is what's stopping me from completely moving to Immich for now.)

Also, I believe this option would be very similar to what PhotoPrism does to sidecars.

**Reply by janhoy** [reply]
Any progress on this? The read-only support is lacking compared to Photoprism so far. We have not yet decided which of the two to go with for my church..

**Comment by shishi** [comment]
This discussion is also a blocker for me from migrating to Immich.I really liked Immich's light and speedy operation, and I prepared for migration by using exiftool to overcome the drawback of it only reading PNG metadata from XMP.
But I considered migrating to Immich and to verify that I would not be locked in, I downloaded and uploaded data using various tools, checking if metadata would be lost or if album information could be maintained, but it was very difficult. I have postponed the migration until the portability of metadata and album information becomes clear.
